Bug fix installs

KZhu-ME opened the pull request:
`installparticlesolver` seems to be working, tested using a dummy print function.
Implemented `installfieldsolver` callback based on `direct_solver` branch, but had to remove `#ifdef AMREX_USE_EB` statements in WarpXEvolve.cpp as the install didn't seem to work when the statements were there.

KZhu-ME committed:
Refactored basestring to str

KZhu-ME committed:
Added installfieldsolver to files, no longer throws errors when called, but does not work

KZhu-ME committed:
Added callfrom function, removed dependency on AMREX_USE_EB

KZhu-ME committed:
Removed AMREX_USE_EB since the fieldsolver function would never trigger

peterscherpelz commented on Source/Evolve/WarpXEvolve.cpp:
```diff
@@ -241,6 +241,7 @@ WarpX::Evolve (int numsteps)
             mypc->SortParticlesByBin(sort_bin_size);
         }
 
+        if (warpx_py_beforeEsolve) warpx_py_beforeEsolve();
```
This needs to be inside the if block, because the Electromagnetic case has already done the Esolve at this point (it's in `OneStep_nosub`, so we only want this for electrostatic.

peterscherpelz commented on Source/Evolve/WarpXEvolve.cpp:
```diff
@@ -249,7 +250,9 @@ WarpX::Evolve (int numsteps)
             bool const reset_fields = true;
-            ComputeSpaceChargeField( reset_fields );
+            if (warpx_py_fieldsolver) warpx_py_fieldsolver();
+                else AddSpaceChargeFieldLabFrame();
```
Indentation needs to be fixed - neither the else nor the line below should be further indented.

peterscherpelz reviewed (changes requested):
I would check before the whole per-step loop in Evolve - like line 41ish - that we didn't install fieldsolver when we're not using an electrostatic solver, as that would be a combination that's very non-intuitive for the user (the fieldsolver doesn't work). eg (wrong syntax, but the idea):
if ( do_electrastic == Ele...Algo::None and fieldsolver_is_installed):
    amrex::Abort("Bad combination!")

We'll need to put more of the embedded boundaries splits @roelof-groenewald made for this to make sense, because this version of the code can't disable the built-in electrostatic solver, which we need to do if a python solver is installed.

I would recommend splitting the beforeEsolve callback insertion into an upstream PR, and we'll hold this PR to include the other EB code that disables built-in-solver too.

KZhu-ME committed:
Added ability to override computePhi(), still needs separate ElectrostaticSolver.cpp file

KZhu-ME committed:
Added back AMREX_USE_EB flag and ability to override computePhi()

KZhu-ME committed:
Added ME_Solver as compile time variable
When ON, enables support for installing custom field solver function in PICMI script

KZhu-ME committed:
Merge branch 'memaster' into bug_fix-installs to get updated fixes from memaster.

KZhu-ME committed:
Changed to check for ME_Solver flag to include ElectrostaticSolver_EB.cpp

KZhu-ME committed:
Changed includes to fix compile errors.

KZhu-ME committed:
Added example for testing installfieldsolver

KZhu-ME committed:
Added ElectrostaticSolver_EB.cpp file

KZhu-ME committed:
Fixed ME_SOLVER flag with GNU make

peterscherpelz commented on Source/FieldSolver/ElectrostaticSolver_EB.cpp:
```diff
@@ -0,0 +1,658 @@
+/* Copyright 2021 Modern Electron
```
Can we move this file to ElectrostaticSolver_ME ? Then it's consistent with the compile var, and we won't get confused when EB is introduced upstream.

peterscherpelz commented on Source/FieldSolver/ElectrostaticSolver_EB.cpp:
```diff
@@ -0,0 +1,658 @@
+#endif
+        }
+    }
+}
```
I believe newline at end of file is standard.

peterscherpelz commented on Source/FieldSolver/Make.package:
```diff
@@ -1,5 +1,9 @@
 CEXE_sources += WarpXPushFieldsEM.cpp
-CEXE_sources += ElectrostaticSolver.cpp
+ifeq ($(ME_Solver),TRUE)
+  CEXE_sources += ElectrostaticSolver_EB.cpp
```
Change name to <same>_ME.cpp here too.

peterscherpelz commented on mewarpx/examples/test_install_fieldsolver.py:
```diff
@@ -0,0 +1,277 @@
+
+# callbacks.installfieldsolver(dummy_solve)
+
+sim.step(5)
```
newline here too

peterscherpelz commented on mewarpx/examples/test_install_fieldsolver.py:
```diff
@@ -0,0 +1,277 @@
+    initial_distribution=uniform_plasma_ion
+)
+
+''' Add back when MCC initialization has been added to mewarpx
```
For this sample that just tests installfieldsolver, eliminate this commented section - and the extraneous stuff at bottom of this script

KZhu-ME committed:
Changed deposit_charge to be true if ME_SOLVER is defined

KZhu-ME committed:
Renamed ElectroStaticSolver_EB.cpp to ElectroStaticSolver_ME.cpp, and fixed code smells

roelof-groenewald commented on Python/pywarpx/callbacks.py:
```diff
@@ -28,6 +28,8 @@
  - beforestep <installbeforestep>: before the time step
  - afterstep <installafterstep>: after the time step
+ - fieldsolver <installfieldsolver>: In place of the AddSpaceChargeFieldLabFrame() call but only in
+                                     an electrostatic simulation with WarpX_EB = ON
```
The flag is no longer WarpX_EB.

roelof-groenewald commented on Source/Evolve/WarpXEvolve.cpp:
```diff
@@ -268,15 +268,30 @@ WarpX::Evolve (int numsteps)
             // positions are next pushed) so that the particles do not deposit out of bounds
             // and so that the fields are at the correct time in the output.
-            bool const reset_fields = true;
+            const bool reset_fields = true;
```
Why this change?

roelof-groenewald commented on Source/Evolve/WarpXEvolve.cpp:
```diff
@@ -268,15 +268,30 @@ WarpX::Evolve (int numsteps)
+#ifdef ME_SOLVER
+            AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
+                do_electrostatic == ElectrostaticSolverAlgo::LabFrame,
+                "Embedded boundary solver can only be used in the labframe."
```
```suggestion
                "ME solver can only be used in the labframe."
```

roelof-groenewald commented on Source/Make.WarpX:
```diff
@@ -99,6 +99,15 @@ ifeq ($(USE_EB),TRUE)
   USERSuffix := $(USERSuffix).EB
 endif
 
+ifeq ($(ME_SOLVER),TRUE)
+  include $(WARPX_HOME)/Source/FieldSolver/Make.package
+  CXXFLAGS += -DME_SOLVER
+  CFLAGS   += -DME_SOLVER
+  FFLAGS   += -DME_SOLVER
+  F90FLAGS += -DME_SOLVER
+  USERSuffix := $(USERSuffix).EB
+endif
+
```
Why do we need this here? The FieldSolver/Make.package file should always be included. Do we actually need the `DME_SOLVER` flags for all those cases? I'm pretty sure the `USERSuffix := $(USERSuffix).EB` line is not needed.

roelof-groenewald reviewed (changes requested):
Left a couple of comments of some initial questions.

KZhu-ME committed:
Refactored code to fix comments and remove redundant changes

KZhu-ME commented:
All the failed azure tests failed when trying to run `python setup.py install --with-libwarpx 2d` in the `Python` directory. This command starts with an error saying that the repository is not a git repository:
```
fatal: not a git repository (or any of the parent directories): .git
```
After giving a lot of warnings, it ends with the error message saying that it can't create or remove files in the install directory:
```
error: can't create or remove files in install directory

The following error occurred while trying to add or remove files in the
installation directory:

    [Errno 13] Permission denied: '/usr/local/lib/python3.6/dist-packages/test-easy-install-17179.write-test'

The installation directory you s pecified (via --install-dir, --prefix, or
the distutils default setting) was:

    /usr/local/lib/python3.6/dist-packages/

Perhaps your account does not have write access to this directory?  If the
installation directory is a system-owned directory, you may need to sign in
as the administrator or "root" account.  If you do not have administrative
access to this machine, you may wish to choose a different installation
directory, preferably one that is listed in your PYTHONPATH environment
variable.

For information on other options, you may wish to consult the
documentation at:

  https://setuptools.readthedocs.io/en/latest/easy_install.html

Please make the appropriate changes for your system and try again.

make: *** [installwarpx] Error 1
```

roelof-groenewald committed:
revert changes to GNUmakefile

roelof-groenewald commented:
@kzhu-ME the Azure test builds still uses the GNUmakefile so we should avoid making any changes to it. The tests failed here (except the Python ones) since the `USE_PYTHON_MAIN` flag was changed to True.

roelof-groenewald commented on Python/pywarpx/callbacks.py:
```diff
@@ -372,6 +378,24 @@ def isinstalledafterdeposition(f):
+    installfieldsolver(f)
+    return f
+def installfieldsolver(f):
+    """Adds a Adds a function to perform the field solve. Note that the C++ object
```
```suggestion
    """Adds a function to perform the field solve. Note that the C++ object
```

roelof-groenewald commented on Source/Evolve/WarpXEvolve.cpp:
```diff
@@ -268,6 +268,7 @@ WarpX::Evolve (int numsteps)
             mypc->SortParticlesByBin(sort_bin_size);
         }
 
+
```
```suggestion
```

roelof-groenewald reviewed (approved):
I left suggestions for quick changes. Otherwise it looks good to me but please remember to address @peterscherpelz's comments about the test input file.

roelof-groenewald commented on Source/FieldSolver/ElectrostaticSolver_ME.cpp:
```diff
@@ -0,0 +1,658 @@
+   \param[in] beta Represents the velocity of the source of `phi`
+*/
+void
+WarpX::computePhi (const amrex::Vector<std::unique_ptr<amrex::MultiFab> >& rho,
+                   amrex::Vector<std::unique_ptr<amrex::MultiFab> >& phi,
+                   std::array<Real, 3> const beta,
+                   Real const required_precision,
+                   int const max_iters) const
+{
+#ifdef WARPX_DIM_RZ
+    computePhiRZ( rho, phi, beta, required_precision, max_iters );
+#else
+    computePhiCartesian( rho, phi, beta, required_precision, max_iters );
+#endif
```
Please also incorporate @MKieburtz's changes here that allows us to set the field solve verbosity.

KZhu-ME committed:
Refactored comment

Co-authored-by: Roelof Groenewald <40245517+roelof-groenewald@users.noreply.github.com>

KZhu-ME committed:
Removed end of file newline

Co-authored-by: Roelof Groenewald <40245517+roelof-groenewald@users.noreply.github.com>

KZhu-ME committed:
Updated ElectroStaticSolver_ME to include verbosity changes

EtherealMC-Bit reviewed (commented):
Sorry for the confusion Kevin but this was the wrong way I went about doing the verbosity first. Here's the right way in this commit and pr. https://github.com/ECP-WarpX/WarpX/pull/2041/commits/31394ab2fdf0eadcec24041e4e30ef71cd308f96

EtherealMC-Bit commented:
> Sorry for the confusion Kevin but this was the wrong way I went about doing the verbosity first. Here's the right way in this commit and pr. [ECP-WarpX@31394ab](https://github.com/ECP-WarpX/WarpX/commit/31394ab2fdf0eadcec24041e4e30ef71cd308f96)

Never mind

nruof commented:
After the verbosity changes that should be good

KZhu-ME committed:
Revert "Updated ElectroStaticSolver_ME to include verbosity changes"

Will wait until verbosity changes are merged with upstream first.
This reverts commit 0c77a4e4dc45943a88c6938ae3b67b7f5f703f07.

KZhu-ME commented:
Is this branch ok to merge now?

peterscherpelz commented:
With two approvals and @MKieburtz 's existing review I'll go ahead and merge. Since there were a lot of commits, I'm going to squash them to keep our history cleaner. We can discuss if we want to make that normal practice, or keep all commits in PRs.
